### ISSUE TITLE
handle_error logic obfuscates original error stack trace

### ISSUE THREAD

**Issue by linkous8** [opsani/servox#79]
If the logic of this function has an error occur, it will prevent any stack trace from showing for the original error: https://github.com/opsani/servox/blob/b4fe0f84ea069a541698dfcb8e6deaf170580da5/servo/connectors/kubernetes.py#L1599

I have a tentative solution here but the resulting logs could still be better organized, just not sure how yet: https://github.com/opsani/servox/blob/d7f8f52e57be5812e6673e91b1bf5d0d824fca48/servo/connectors/kubernetes.py#L2052-L2078

Note [example_output.txt](https://github.com/opsani/servox/files/5525663/example_output.txt) and how we only see the stack trace for the failed rollback and very minimal info on the original error (422
Reason: Unprocessable Entity)

**Comment by blakewatters** [comment]
This is a good catch and I think there's a simpler solution:

We can use the `raise [ErrorType] from error` explicit exception chaining syntax (see [PEP 3134 -- Exception Chaining and Embedded Tracebacks](https://www.python.org/dev/peps/pep-3134/#explicit-exception-chaining)) to wrap the underlying error and ensure that the traceback is maintained.

Does this change cover it without having to trampoline through the try/except block?

```python
async def handle_error(self, error: Exception, mode: "FailureMode") -> bool:
    if mode == FailureMode.CRASH:
        raise RuntimeError(
            "an unrecoverable failure occurred while interacting with Kubernetes"
        ) from error

    elif mode == FailureMode.IGNORE:
        self.logger.warning(f"ignoring runtime error and continuing: {error}")
        self.logger.opt(exception=error).exception("ignoring Kubernetes error")
        return True

    elif mode == FailureMode.ROLLBACK:
        await self.rollback(error)
        return True

    elif mode == FailureMode.DESTROY:
        await self.destroy(error)
        return True

    else:
        raise NotImplementedError(
            f"missing error handler for failure mode '{mode}'"
        ) from error
```

**Comment by blakewatters** [comment]
Went ahead and pushed the error chaining to master. Lemme know if that covers it or we need to do more

**Comment by linkous8** [comment]
The change is good, it covers one of the cracks the original error can slip through. However, it does not address particular crack I raised this issue for. Note in the following stack trace that an exception is bubbling out of the call to `await self.rollback(error)` causing the original error to be discarded:

```
  File "/home/linkous/code_review/servox/servo/connectors/kubernetes.py", line 2309, in apply
    if await optimization.handle_error(
             │            └ <function BaseOptimization.handle_error at 0x7fc446130dc0>
             └ DeploymentOptimization(name='web-main/main', timeout=Duration('5m' 0:05:00), deployment_config=DeploymentConfiguration(descri...

  File "/home/linkous/code_review/servox/servo/connectors/kubernetes.py", line 1630, in handle_error
    await self.rollback(error)
          │    │        └ ApiException()
          │    └ <function DeploymentOptimization.rollback at 0x7fc44613b310>
          └ DeploymentOptimization(name='web-main/main', timeout=Duration('5m' 0:05:00), deployment_config=DeploymentConfiguration(descri...

  File "/home/linkous/code_review/servox/servo/connectors/kubernetes.py", line 1762, in rollback
    await asyncio.wait_for(
          │       └ <function wait_for at 0x7fc448505f70>
          └ <module 'asyncio' from '/home/linkous/.pyenv/versions/3.8.5/lib/python3.8/asyncio/__init__.py'>

  File "/home/linkous/.pyenv/versions/3.8.5/lib/python3.8/asyncio/tasks.py", line 483, in wait_for
    return fut.result()
           │   └ <method 'result' of '_asyncio.Future' objects>
           └ <_GatheringFuture finished exception=ApiException()>

  File "/home/linkous/code_review/servox/servo/connectors/kubernetes.py", line 1177, in rollback
    self.obj = await api_client.create_namespaced_deployment_rollback(
    │    │           │          └ <function ExtensionsV1beta1Api.create_namespaced_deployment_rollback at 0x7fc4463360d0>
    │    │           └ <kubernetes_asyncio.client.api.extensions_v1beta1_api.ExtensionsV1beta1Api object at 0x7fc443d3f430>
    │    └ {'api_version': 'apps/v1',
    │       'kind': 'Deployment',
    │       'metadata': {'annotations': {'deployment.kubernetes.io/revision': '1',
    │         ...
    └ {'api_version': 'apps/v1',
       'kind': 'Deployment',
       'metadata': {'annotations': {'deployment.kubernetes.io/revision': '1',
         ...

  File "/home/linkous/.cache/pypoetry/virtualenvs/servox-eNeZ9xha-py3.8/lib/python3.8/site-packages/kubernetes_asyncio/client/api_client.py", line 189, in __call_api
    raise e
  File "/home/linkous/.cache/pypoetry/virtualenvs/servox-eNeZ9xha-py3.8/lib/python3.8/site-packages/kubernetes_asyncio/client/api_client.py", line 182, in __call_api
    response_data = await self.request(
                          │    └ <function ApiClient.request at 0x7fc446404940>
                          └ <kubernetes_asyncio.client.api_client.ApiClient object at 0x7fc443d3f640>
  File "/home/linkous/.cache/pypoetry/virtualenvs/servox-eNeZ9xha-py3.8/lib/python3.8/site-packages/kubernetes_asyncio/client/rest.py", line 230, in POST
    return (await self.request("POST", url,
                  │    │               └ 'https://192.168.39.227:8443/apis/extensions/v1beta1/namespaces/default/deployments/web-main/rollback'
                  │    └ <function RESTClientObject.request at 0x7fc4463f9e50>
                  └ <kubernetes_asyncio.client.rest.RESTClientObject object at 0x7fc443d3fe50>
  File "/home/linkous/.cache/pypoetry/virtualenvs/servox-eNeZ9xha-py3.8/lib/python3.8/site-packages/kubernetes_asyncio/client/rest.py", line 187, in request
    raise ApiException(http_resp=r)
          │                      └ <kubernetes_asyncio.client.rest.RESTResponse object at 0x7fc445e681f0>
          └ <class 'kubernetes_asyncio.client.exceptions.ApiException'>

kubernetes_asyncio.client.exceptions.ApiException: (404)
Reason: Not Found
HTTP response headers: <CIMultiDictProxy('Cache-Control': 'no-cache, private', 'Content-Type': 'application/json', 'Date': 'Wed, 11 Nov 2020 17:13:31 GMT', 'Content-Length': '174')>
HTTP response body: {"kind":"Status","apiVersion":"v1","metadata":{},"status":"Failure","message":"the server could not find the requested resource","reason":"NotFound","details":{},"code":404}
```

My proposed solution is to wrap the handle_error() logic in a try/catch block that ensures both exceptions get logged/raised like such (I'm not sure which error should raise from which):

https://github.com/opsani/servox/blob/a6c82478022d1f1d8604896b7b3463cc8d7b41a9/servo/connectors/kubernetes.py#L2146-L2173

**Comment by blakewatters** [comment]
I see your point here. Thanks for following up and reopening.

In [ca4b3e5](https://github.com/opsani/servox/commit/ca4b3e5a20baa58921503d4845058fbe22863acc) I just did the same trick again using the `raise from` syntax.

We don't need to trap them separately and log/reraise independently. When PEP 3134 was adopted the traceback module and friends were updated to become aware of exception chaining. Covering this other missed chaining scenario should mean that whatever catches it downstream will show you the full chain.

Reopen if I again missed something.
